File: java/ReorderList.java
/*
 * Filename: WordBreak.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 03 Jul 2015 09:58:52 AM EDT
 */

public class Solution {
	public void reorderList(ListNode head) {
		if (head == null || head.next == null || head.next.next == null) {
			return;
		}

		ListNode p1 = head;
		ListNode p2 = head;

		while (p1.next != null && p1.next.next != null) {
			p1 = p1.next.next;
			p2 = p2.next;
		}

		ListNode head2 = p2.next;
		p2.next = null;
		head2 = reverse(head2);
		merge(head, head2);
	}

	private ListNode reverse(ListNode head) {
		ListNode tail = null;
		ListNode tmp = head;

		while (tmp != null) {
			head = tmp;
			tmp = tmp.next;
			head.next = tail;
			tail = head;
		}

		return head;
	}

	private void merge(ListNode head, ListNode head2) {
		ListNode tmp;

		while (head2 != null) {
			tmp = head2.next;
			head2.next = head.next;
			head.next = head2;
			head = head.next.next;
			head2 = tmp;
		}
	}
}

File: cpp/reorderList.cpp
/*
 * Filename: reorderList.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Mon 18 Jan 2016 07:55:35 PM EST
 * Description: 
 */

#include <iostream>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution
{
	public: void reorderList(ListNode* head)
		{
			if (!head || !head->next)
			{
				return;
			}

			ListNode* p = head;
			ListNode* q = head;

			while (p->next && p->next->next)
			{
				p = p->next->next;
				q = q->next;
			}

			ListNode* head2 = q->next;
			q->next = NULL;
			head2 = reverse(head2);
			merge(head, head2);
		}

	private:
		ListNode* reverse(ListNode* head)
		{
			ListNode* tail = NULL;
			ListNode* tmp;

			while (head)
			{
				tmp = head->next;
				head->next = tail;
				tail = head;
				head = tmp;
			}

			return tail;
		}

		void merge(ListNode* head, ListNode* head2)
		{
			ListNode* tmp;

			while (head2)
			{
				tmp = head->next;
				head->next = head2;
				head2 = head2->next;
				head = head->next;
				head->next = tmp;
				head = head->next;
			}
		}
};
